### PR TITLE
Document dotnet-gcdump tool

### LINKED ISSUE #18935: Missing docs for dotnet-gcdump
Either a new page should be created, or it should be refererenced on this page.

See https://devblogs.microsoft.com/dotnet/collecting-and-analyzing-memory-dumps/
for more info.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9a269a64-0a27-27d3-a509-47b9a0d114bf
* Version Independent ID: 49ff97ff-b39a-5ae8-e4b8-81a4f8eab50c
* Content: [dotnet-dump - .NET Core](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump)
* Content Source: [docs/core/diagnostics/dotnet-dump.md](https://github.com/dotnet/docs/blob/master/docs/core/diagnostics/dotnet-dump.md)
* Product: **dotnet-core**
* GitHub Login: @sdmaclea
* Microsoft Alias: **stmaclea**

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #18935

[Information source](https://github.com/dotnet/diagnostics/blob/f63dc82eb4d92b8e057f7e3b5855cdf06f985f8c/documentation/dotnet-gcdump-instructions.md).

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-gcdump?branch=pr-en-us-19697).

### CODE REVIEW

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:6]: Following the pattern of the other docs:
```suggestion
# dotnet-gcdump
```

Also it's best to avoid applying text styles to headings.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:10]:
```suggestion
## Install dotnet-gcdump
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:42]: A Commands section here like the one in https://raw.githubusercontent.com/dotnet/docs/master/docs/core/diagnostics/dotnet-counters.md might be helpful. Then the following h2s would correspond to the commands in the table.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:80]:
```suggestion
Lists the dotnet processes that dumps can be collected for.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:52]: Following the pattern of https://raw.githubusercontent.com/dotnet/docs/master/docs/core/diagnostics/dotnet-counters.md we should have an  Examples h3 section following each Options h3 section.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:26]: gcdumps/gcdumps isn't a word
```suggestion
The `dotnet-gcdump` global tool is a way to collect dumps of live .NET processes. It uses the EventPipe technology, which is a cross-platform alternative to ETW on Windows. Dumps are created by triggering a garbage collection (GC) in the target process, turning on special events, and regenerating the graph of object roots from the event stream. This allows for dumps to be collected while the process is running, with minimal overhead. These dumps are useful for several scenarios:
```
Can EventPipe technology be a link to an overview of that?

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:30]:
```suggestion
- Comparing the number of objects on the heap at several points in time.
- Analyzing roots of objects (answering questions like, "what still has a reference to this type?").
- Collecting general statistics about the counts of objects on the heap.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:44]:
```suggestion
Collects a dump from a currently running process.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:60]:
```suggestion
  The process id to collect the dump.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:64]:
```suggestion
  The path where collected dumps should be written. Defaults to *.\\YYYYMMDD\_HHMMSS\_\<pid>.gcdump*.
```
An example would be helpful.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:68]:
```suggestion
  Output the log while collecting the dump.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:72]:
```suggestion
  Give up on collecting the dump if it takes longer than this many seconds. The default value is 30.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:76]:
```suggestion
  The name of the process to collect the dump for.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:90]:
```suggestion
Generate a report from a previously generated dump or from a running process, and write to `stdout`.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:106]:
```suggestion
  The process id to collect the dump for.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:110]: If this is the only option, what's the difference between specifying it an omitting it?

**Review by tdykstra — COMMENTED** [review]
This is a good start -- I had a few doc structure and wording suggestions.  I would put this in the TOC following dotnet-dump

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:6]: The current heading is consistent with https://github.com/dotnet/docs/blob/master/docs/core/diagnostics/dotnet-dump.md
I wouldn't mind remove the text style from both the headings, but I personally like keeping the small description of the tool in the H1s. It also recommended that H1s be different from titles AFAIK.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:80]: @tdykstra why change gcdumps to dumps?
The "for" word seems to be missing in the tool itself too.

I opened https://github.com/dotnet/diagnostics/pull/1387 to fix that in the tool.

I'll apply adding "for" now, and wait for your feedback about changing gcdumps to dumps

```suggestion
Lists the dotnet processes that gcdumps can be collected for.
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:110]: @tdykstra, I had this question myself. It looks like there shouldn't be a difference. Perhaps the tool team might be planning to add other options later?

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:42]: @tdykstra, The live docs has a side-bar containing links to all H2s, I personally think a commands section with only links to the heading is noisy and repetitive with the sidebar.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:26]: Would that make any confusion with the `dotnet-dump` tool?

I'm not familiar with any of both the tools, but I think instead of replacing gcdumps with dumps, we may change it to gc dumps or GC dumps?

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:106]: Is "trace" not correct here?
This is exactly what `dotnet-gcdump report -h` prints.

image

If trace is incorrect or inaccurate, the tool should be updated.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [reply docs/core/diagnostics/dotnet-gcdump.md:6]: >keeping the small description of the tool in the H1s

OK

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [reply docs/core/diagnostics/dotnet-gcdump.md:26]: The gcdump tool produces dumps -- in context there should be no cause for confusion because we're not talking about the dump tool. That said, if you prefer to use GC dumps I'm OK with that as long as at the first occurrence of GC the acronym is spelled out.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [reply docs/core/diagnostics/dotnet-gcdump.md:42]: OK to omit the Commands section.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [reply docs/core/diagnostics/dotnet-gcdump.md:106]: my guess is that's a copy/paste error  from dotnet-trace.

But looking at the dotnet-trace doc I see it uses "from" where I put "for" at the end of sentences like "... to collect the dump from" -- for consistency I'd go with "from" instead of "for" in this doc.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:6]:
```suggestion
# Heap analysis tool (dotnet-gcdump)
```

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:106]: I've opened https://github.com/dotnet/diagnostics/pull/1388 to fix that in the tool side.

```suggestion
  The process id to collect the GC dump from.
```

**Comment by Youssef1313** [comment]
The options here are in the same order produced by the CLI help, do you want me to order them alphabetically as done with dotnet-trace? @tdykstra

**Comment by tdykstra** [comment]
>The options here are in the same order produced by the CLI help, do you want me to order them alphabetically as done with dotnet-trace?

There's no reason the docs have to follow the CLI help order; any way we can make the docs more usable seems worth doing to me. In this case the small number of options makes the benefit of limited value, but my preference is to be consistent.

**Review by tdykstra — COMMENTED** [review]
It doesn't look like the description section incorporates the info from https://github.com/dotnet/diagnostics/blob/f63dc82eb4d92b8e057f7e3b5855cdf06f985f8c/documentation/dotnet-gcdump-instructions.md#known-caveats and https://github.com/dotnet/diagnostics/blob/f63dc82eb4d92b8e057f7e3b5855cdf06f985f8c/documentation/dotnet-gcdump-instructions.md#viewing-the-gcdump-captured-from-dotnet-gcdump -- is that deliberate?

**Comment by Youssef1313** [comment]
> deliberate

I'll add these. Thanks for noticing this.

File: docs/core/additional-tools/index.md
Comment (pkulikov) [inline docs/core/additional-tools/index.md:22]:
```suggestion
[dotnet-gcdump](../diagnostics/dotnet-gcdump.md) provides a way to collect GC (Garbage Collector) dumps of live .NET processes.
```

File: docs/core/diagnostics/index.md
Comment (Youssef1313) [inline docs/core/diagnostics/index.md:38]:
```suggestion
The [dotnet-gcdump](dotnet-gcdump.md) tool is a way to collect GC (Garbage Collector) dumps of live .NET processes.
```

File: docs/core/additional-tools/index.md
Comment (Youssef1313) [reply docs/core/additional-tools/index.md:22]: @pkulikov, Thanks. I missed this.

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (sdmaclea) [reply docs/core/diagnostics/dotnet-gcdump.md:110]: /cc @josalem

Comment (josalem) [reply docs/core/diagnostics/dotnet-gcdump.md:110]: This is only report type available right now and therefore the default, but we created this flag for future use if/when we add more report types.

**Review by josalem — APPROVED** [review]

**Comment by adegeo** [comment]
@tdykstra Is this good to go or do you have other concerns?

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:112]: This might fit as an H3 in the Description section

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:114]: mdlint error -- could make PerfView a link

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (tdykstra) [inline docs/core/diagnostics/dotnet-gcdump.md:118]:
```suggestion
## Troubleshoot
```

**Review by tdykstra — COMMENTED** [review]
A couple suggestions

File: docs/core/diagnostics/dotnet-gcdump.md
Comment (Youssef1313) [reply docs/core/diagnostics/dotnet-gcdump.md:114]:
```suggestion
On Windows, `.gcdump` files can be viewed in [PerfView](https://github.com/microsoft/perfview) for analysis or in Visual Studio. Currently, There is no way of opening a `.gcdump` on non-Windows platforms.
```

**Comment by tdykstra** [comment]
There's still a markdownlint error - we can merge this when that's resolved. 

>The PR still lacks examples. I'll try adding some.

You can save that for a later PR if you prefer.

**Comment by Youssef1313** [comment]
> 
> 
> There's still a markdownlint error - we can merge this when that's resolved.
> 
> > The PR still lacks examples. I'll try adding some.
> 
> You can save that for a later PR if you prefer.

I completely forgot about that. If it's okay to merge this, I'll open an issue to track adding examples. Otherwise, I'll find time to add them tomorrow or so.

**Comment by tdykstra** [comment]
Yes OK to merge and add later.  Thanks for creating this doc!

**Pull request merged**
